### PR TITLE
Added liquibase:snapshot command to the liquibase-maven-plugin

### PR DESCRIPTION
**Pull request by malikin**
- - -
name: Pull Request
about: Create a report to help us improve
title: Added liquibase:snapshot command to the liquibase-maven-plugin
labels: Status:Discovery
assignees: malikin

- - -
## Environment
**Liquibase Version**: 4.0.1

**Liquibase Integration & Version**:

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**:

**Operating System Type & Version**:

## Pull Request Type
Added liquibase snapshot command to liquibase-maven-plugin. `mvn liquibase:snapshot`

* [ ] Bug fix (non-breaking change which fixes an issue.)
* [x] Enhancement/New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
I have found that maven plugin does not support `snapshot` command. This PR added the command. https://docs.liquibase.com/commands/community/snapshot.html

## Steps To Reproduce
`mvn liquibase:snapshot`

## Actual Behavior
`[ERROR] Could not find goal 'snapshot' in plugin org.liquibase:liquibase-maven-plugin:4.0.0 among available goals changelogSync, changelogSyncSQL, clearCheckSums, dbDoc, diff, dropAll, futureRollbackSQL, generateChangeLog, help, history, listLocks, releaseLocks, rollback, rollbackOneChangeSet, rollbackOneChangeSetSQL, rollbackOneUpdate, rollbackOneUpdateSQL, rollbackSQL, status, tag, update, updateSQL, updateTestingRollback -> [Help 1]
org.apache.maven.plugin.MojoNotFoundException: Could not find goal 'snapshot' in plugin org.liquibase:liquibase-maven-plugin:4.0.0 among available goals changelogSync, changelogSyncSQL, clearCheckSums, dbDoc, diff, dropAll, futureRollbackSQL, generateChangeLog, help, history, listLocks, releaseLocks, rollback, rollbackOneChangeSet, rollbackOneChangeSetSQL, rollbackOneUpdate, rollbackOneUpdateSQL, rollbackSQL, status, tag, update, updateSQL, updateTestingRollback`

## Expected/Desired Behavior
After `mvn liquibase:snapshot` it will make snapshot to `outputFile` or stdout like here https://docs.liquibase.com/commands/community/snapshot.html

## Fast Track PR Acceptance Checklist:

* [ ] Build is successful and all new and existing tests pass
* [ ] Added [Unit Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
* [ ] Added [Integration Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)
* [ ] Documentation Updated

## Need Help?
Come chat with us on our [discord channel](https://discord.com/channels/700506481111597066/700506481572839505)

### CODE REVIEW

**Comment by filipelautert** [comment]
Funny that no one commented or reviewed this one.

**Comment by MalloD12** [comment]
Hi @malikin,

Thank you for submitting this PR. It's good to have now a `snapshot` command implementation for `maven` plugin. The `performLiquibaseTask()` method implementation was not totally accurate because it was using `SnapshotCommand` class which is deprecated and invoking a `execute()` method that wasn't available. I updated this by using the CommandStep framework, please have a look at it to see if you are happy with these changes. If not feel free to modify it.

I manually test snapshot command and see it is working.   

Thanks,
Daniel.

**Comment by malikin** [comment]
@MalloD12 Hi, thank you so much, it's been a while since I created this PR, almost forgot about it :) 
Everything looks good to me!

**Review by MalloD12 — APPROVED** [review]
**Approved**.

Most of the code changes provided by @malikin looks good to me. I only made an update to the `performLiquibaseTask` implementation to avoid using `SnapshotCommand` class (which is deprecated) and use `SnapshotCommandStep` class instead.

This changes have been manually tested by me, but it would nice if @rberezen or @tati-qalified can perform any additional testing. 

Thanks,
Daniel.

File: liquibase-maven-plugin/src/main/java/org/liquibase/maven/plugins/LiquibaseSnapshotMojo.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+     * Specifies the file path to where the snapshot JSON or YAML will be written.
+     *
+     * @parameter property="liquibase.outputFile"
+     */
+    protected File outputFile;
+
+    /**
+     * The writer for writing the snapshot file.
+     */
+    private Writer outputWriter;
```
Comment (filipelautert) [inline liquibase-maven-plugin/src/main/java/org/liquibase/maven/plugins/LiquibaseSnapshotMojo.java:42]: This one should be a local variable.

**Review by filipelautert — CHANGES_REQUESTED** [review]
Sonar is failing for this PR -> https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=1379&id=liquibase .

File: liquibase-maven-plugin/src/main/java/org/liquibase/maven/plugins/LiquibaseSnapshotMojo.java (hunk 1 above)
Comment (malikin) [reply liquibase-maven-plugin/src/main/java/org/liquibase/maven/plugins/LiquibaseSnapshotMojo.java:42]: Thanks, I'll fix this later today and SonarQube issues

Comment (MalloD12) [reply liquibase-maven-plugin/src/main/java/org/liquibase/maven/plugins/LiquibaseSnapshotMojo.java:42]: Thank you, @malikin. Please let me know if you need any help!

Comment (malikin) [reply liquibase-maven-plugin/src/main/java/org/liquibase/maven/plugins/LiquibaseSnapshotMojo.java]: We don't need it actually, with the "CommandStep framework" code is much simpler. I removed the unused method and this property too, should be good now.

Comment (filipelautert) [reply liquibase-maven-plugin/src/main/java/org/liquibase/maven/plugins/LiquibaseSnapshotMojo.java]: Thanks!

**Comment by MalloD12** [comment]
Hi @malikin,

@tati-qalified has been testing this PR today and and found a NPE when `outputFile` is not speicified:
```
Caused by: java.lang.NullPointerException
    at java.io.FileOutputStream.<init> (FileOutputStream.java:226)
    at java.io.FileOutputStream.<init> (FileOutputStream.java:184)
    at org.liquibase.maven.plugins.LiquibaseSnapshotMojo.performLiquibaseTask (LiquibaseSnapshotMojo.java:47)
```
We need to add a null check when `outputFile` is not specified. This error was not happening before because this logic was handled in `createLiquibase()` method. Could you please fix that?

Thanks,
Daniel.

**Comment by malikin** [comment]
@MalloD12 Thanks, sure, will fix that!

**Comment by malikin** [comment]
@MalloD12 NPE is fixed now, if `outpuFile` parameter is not set, the snapshot data goes to `System.out` (like in `LiquibaseDatabaseDiff`). I also replaced `FileOutputStream` with `PrintStream` - never throws an `IOException`, the flush method is automatically invoked, etc.

**Comment by MalloD12** [comment]
@tati-qalified I think you are good to resume your testing. I tried without specifying an output file and `snapshot` was executed successfully.

Thank you, @malikin!

**Comment by tati-qalified** [comment]
@malikin in general this looks great! There are just two things worth mentioning:

1. By default it's outputting a JSON format, but it should be a text. Using `-Dliquibase.snapshotFormat=txt` gives the correct output.
2. The snapshot ids differ between the CLI and the plugin. Here's an example:
image

**Comment by malikin** [comment]
@tati-qalified Thank you so much for testing!

> 1. By default it's outputting a JSON format, but it should be a text.
Ok, I can update this

> 2. The snapshot ids differ between the CLI and the plugin. 
I'm pretty sure this is ok, an `id` is uniq for each run - if you try to run serveral times in a row, I'll get a different `id` for each run.

**Comment by malikin** [comment]
@tati-qalified I just tried to run the  `./mvnw clean compile liquibase:snapshot` on my test project with `liquibase-maven-plugin.properties` without `snapshotFormat` property, and it generated `txt` snapshot. 

Also, this line confirms that `SnapshotCommandStep` uses `txt` by default, my code calls it https://github.com/liquibase/liquibase/blob/master/liquibase-standard/src/main/java/liquibase/command/core/SnapshotCommandStep.java#L118

I updated the parameter description, it was confusing (I wrote it almost 4 years ago, I suppose it was a `json` by default back than :)

**Comment by tati-qalified** [comment]
@malikin hmm... that's strange. I'll test it again, maybe it was some mistake on my part

**Comment by tati-qalified** [comment]
@malikin I'm still getting a JSON format. Are you using Windows or a different OS?

I'm using Windows, and here are my configurations and the command I'm running:
```
<configuration>
      <changeLogFile>dbchangelogs/master-changelog.xml</changeLogFile>
      <url>jdbc:h2:tcp://localhost:9090/mem:dev</url>
      <username>dbuser</username>
      <password>letmein</password>
</configuration>
```

`mvn liquibase:snapshot`


Note that there is no specified format anywhere, nor any output file that might change the resulting format.
The output starts like this:
```
{
  "snapshot": {
    "created": "2024-07-09T11:20:08.521",
    "database": {
      "productVersion": "2.2.224 (2023-09-17)",
      "shortName": "h2",
      "majorVersion": "2",
      "minorVersion": "2",
      "user": "DBUSER",
      "productName": "H2",
      "url": "jdbc:h2:tcp://localhost:9090/mem:dev"
    },

```
so it's clearly JSON.

I think the OS might be having some effect on this, but if not then I'm not sure. Any ideas?
CC @MalloD12

**Comment by tati-qalified** [comment]
@malikin the configuration does say json...
```
----------------------------
[INFO]     driver: null
[INFO]     url: jdbc:h2:tcp://localhost:9090/mem:dev
[INFO]     username: *****
[INFO]     password: *****
[INFO]     use empty password: false
[INFO]     properties file: null
[INFO]     properties file will override? false
[INFO]     clear checksums? false
[INFO]     changeLogDirectory: null
[INFO]     changeLogFile: dbchangelogs/master-changelog.xml
[INFO]     context(s): null
[INFO]     label(s): null
[INFO]     snapshotFormat: json
[INFO]     outputFile: null
[INFO] ------------------------------------------------------------------------
```

but I haven't set that format anywhere. I've checked my environment variables as well, and it's not there.

I'm honestly confused, I've done the usual "turn it off and on again", recompile, etc., so I don't know why it's not working. If others can't replicate it, I'd say that we move forward with this and just leave a note that this may happen. It's an easy workaround anyway.

@filipelautert could you try to replicate it?

Thanks!

**Review by tati-qalified — APPROVED** [review]

**Comment by MalloD12** [comment]
Hey guys, looking at the discussion you both were having I also tried running: `mvn clean compile liquibase:snapshot` and I can see this on my output:

```
 <snapshotFormat default-value="json">${liquibase.snapshotFormat}</snapshotFormat>
```

and my setup looks like:

```
 <configuration>
        <propertyFile>liquibase.workspace.postgres.properties</propertyFile>
         <diffTypes>table,column,data,view</diffTypes>
</configuration>
```

and `liquibase.workspace.postgres.properties` content like:

```
changeLogFile= changelog/xml/xsd-test.xml
username=postgres
password=mypass
url: jdbc:postgresql://localhost:5438/lbcat
```

**Comment by malikin** [comment]
@MalloD12 did you pull the very last version from the branch?

**Comment by MalloD12** [comment]
Hi @malikin,

Yeah, my bad I was not trying with the latest changes from this branch. I can see now that the default output is in `txt` format.  Although, I was telling @tati-qalified I cannot see this section:

```
[INFO] Settings
----------------------------
[INFO]     driver: null
[INFO]     url: jdbc:mysql://localhost:3306/sakila
[INFO]     username: *****
[INFO]     password: *****
[INFO]     use empty password: false
[INFO]     properties file: null
[INFO]     properties file will override? false
[INFO]     clear checksums? false
[INFO]     changeLogDirectory: null
[INFO]     changeLogFile: /src/main/resources/db/changelog/db.changelog-master.yaml
[INFO]     context(s): null
[INFO]     label(s): null
[INFO]     snapshotFormat: txt
[INFO]     outputFile: null
```

Thanks,
Daniel.

**Review by rberezen — APPROVED** [review]

**Comment by MalloD12** [comment]
@AMBERMW13  - @adrian-velonis1: guys, I've created this PD [ticket](https://datical.atlassian.net/browse/PD-5348) to document this change.

Thanks,
Daniel.

**Pull request merged**